patches/minecraft/net/minecraft/tags/TagLoader.java.patch

```diff
@@ -1,11 +1,108 @@
+-                     list.add(new TagLoader.EntryWithSource(p_215997_, s1));
++                     // FORGE: Specify this tag is being added
++                     list.add(new TagLoader.EntryWithSource(p_215997_, s1, true));
++                  });
++                  // FORGE: Add removed tags
++                  tagfile.removeEntries().forEach(entry1 -> {
++                     list.add(new TagLoader.EntryWithSource(entry1, s1, false));
+                   });
+                } catch (Throwable throwable1) {
+                   if (reader != null) {
+@@ -119,21 +_,30 @@
+ 
+    }
+ 
++   /** @deprecated Forge: Use {{@link #buildPair(net.minecraft.tags.TagEntry.Lookup, List)} instead. */
++   @Deprecated
+    private Either<Collection<TagLoader.EntryWithSource>, Collection<T>> m_215978_(TagEntry.Lookup<T> p_215979_, List<TagLoader.EntryWithSource> p_215980_) {
+-      ImmutableSet.Builder<T> builder = ImmutableSet.builder();
++      return buildSets(p_215979_, p_215980_).mapRight(p -> p.additive());
++   }
++
++   // FORGE: Build both the additive and subtractive sets for the given tags
++   private Either<Collection<TagLoader.EntryWithSource>, AdditiveSubtractiveSets<T>> buildSets(TagEntry.Lookup<T> lookup, List<TagLoader.EntryWithSource> entries) {
++      ImmutableSet.Builder<T> additive = ImmutableSet.builder();
++      ImmutableSet.Builder<T> subtractive = ImmutableSet.builder();
+       List<TagLoader.EntryWithSource> list = new ArrayList<>();
+ 
+-      for(TagLoader.EntryWithSource tagloader$entrywithsource : p_215980_) {
+-         if (!tagloader$entrywithsource.f_216042_().m_215927_(p_215979_, builder::add)) {
++      for(TagLoader.EntryWithSource tagloader$entrywithsource : entries) {
++         if (!tagloader$entrywithsource.f_216042_().m_215927_(lookup, tagloader$entrywithsource.additive() ? additive::add : subtractive::add)) {
+             list.add(tagloader$entrywithsource);
+          }
+       }
+ 
+-      return list.isEmpty() ? Either.right(builder.build()) : Either.left(list);
++      return list.isEmpty() ? Either.right(new AdditiveSubtractiveSets<>(additive.build(), subtractive.build())) : Either.left(list);
+    }
+ 
+    public Map<ResourceLocation, Collection<T>> m_203898_(Map<ResourceLocation, List<TagLoader.EntryWithSource>> p_203899_) {
+-      final Map<ResourceLocation, Collection<T>> map = Maps.newHashMap();
++      // FORGE: Store the full information provided from the private build method
++      final Map<ResourceLocation, AdditiveSubtractiveSets<T>> map = Maps.newHashMap();
+       TagEntry.Lookup<T> lookup = new TagEntry.Lookup<T>() {
+          @Nullable
+          public T m_213619_(ResourceLocation p_216039_) {
+@@ -142,7 +_,9 @@
+ 
+          @Nullable
+          public Collection<T> m_214048_(ResourceLocation p_216041_) {
+-            return map.get(p_216041_);
++            // FORGE: Only return the additive sets
++            if (map.get(p_216041_) == null) { return null; }
```

Comment:
Brackets can be removed

Reply:
I could have sworn I did this, but I guess not...